lib/mayaUsd/ufe/UsdMaterial.cpp

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright 2022 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdMaterial.h"
+
+#include <mayaUsd/ufe/Utils.h>
+
+#include <ufe/types.h>
+
+#include <stdexcept>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+UsdMaterial::UsdMaterial(const UsdSceneItem::Ptr& item)
+    : Ufe::Material()
+    , fItem(item)
+{
+}
+
+UsdMaterial::~UsdMaterial() { }
+
+/*static*/
+UsdMaterial::Ptr UsdMaterial::create(const UsdSceneItem::Ptr& item)
+{
+    return std::make_shared<UsdMaterial>(item);
+}
+
+std::vector<PXR_NS::UsdPrim> UsdMaterial::getConnectedShaders(const PXR_NS::UsdShadeOutput& port) const
+{
+    // Dig down across NodeGraph boundaries until a surface shader is found.
+    std::vector<PXR_NS::UsdPrim> shaders;
+    for (const auto& sourceInfo : port.GetConnectedSources()) {
+        if (sourceInfo.source.GetPrim().GetTypeName() != "Shader") {
+            if (sourceInfo.sourceType == UsdShadeAttributeType::Output) {
+                const std::vector<PXR_NS::UsdPrim> connectedShaders
+                    = getConnectedShaders(sourceInfo.source.GetOutput(sourceInfo.sourceName));
+                shaders.insert(end(shaders), begin(connectedShaders), end(connectedShaders));
```

Comment:
For `end` and `begin` the code should be `std::end` and `std::begin`. Why the compilation succeeds?

Reply:
Looks like a case of [Argument-dependent Lookup](https://en.cppreference.com/w/cpp/language/adl). The namespace of std::map::insert ca be searched for begin/end functions.